docs/reference/programming_model.rst

```diff
@@ -0,0 +1,259 @@
+.. meta::
+  :description: This chapter describes the HIP programming model, the contract
+                between the programmer and the compiler/runtime executing the
+                code.
+  :keywords: AMD, ROCm, HIP, CUDA, C++ language extensions
+
+*******************************************************************************
+Programming Model
+*******************************************************************************
+
+HIP defines a model of mapping SIMT programs (Single Instruction, Multiple
+Threads) onto various architectures, primarily GPUs. While the model may be
+expressed in most imperative languages, (eg. Python via PyHIP) this document
+will focus on the original C/C++ API of HIP.
+
+Threading Model
+===============================================================================
+
+The SIMT nature of HIP is captured by the ability to execute user-provided
+device programs, expressed as single-source C/C++ functions or sources compiled
+online/offline to binaries in bulk.
+
+Multiple instances of the device program (aka. kernel) may execute in parallel,
+all uniquely identified by a set of integral values which are referred to as
+thread IDs. The set of integers identifying a thread relate to the hierarchy in
+which threads execute.
+
+.. _inherent_thread_model:
+
+Inherent Thread Model
+-------------------------------------------------------------------------------
+
+The thread hiearchy inherent to how AMD GPUs operate manifest as depicted in
+:numref:`inherent_thread_hierarchy`.
+
+.. _inherent_thread_hierarchy:
+
+.. figure:: ../data/reference/programming_model/thread_hierarchy.svg
```

Comment:
While I do agree with the sentiment of not conflating the physical realization from the logical structure, I don't agree with warps not being part of the API. Warp-collective functions explicitly operate at this granularity, and if the notion isn't introduced, it can't be referred to by those functions.

A block consisting of warps where the last has unnecessary threads is a valid concern though. We may mention that in text. Highlighting that capability in image may be too much in the opposite direction: encouraging using the API as such.

Reply:
Added a sentence mentioning threads, at least. I'll open another issue to keep track of this - this image should be updated